File: java/step2-early-return/src/main/java/badcode/DeepRaceValidator.java
package badcode;

public class DeepRaceValidator {
    private static final int MAX_NAME_LENGTH = 5;
    private static final int MOVE_THRESHOLD = 4;
    private static final int MAX_RANDOM_VALUE = 9;

    public int move(String name, int randomValue, int currentPosition) {
        if (name != null) {
            if (!name.isEmpty()) {
                if (name.length() <= MAX_NAME_LENGTH) {
                    if (randomValue >= 0) {
                        if (randomValue <= MAX_RANDOM_VALUE) {
                            if (randomValue >= MOVE_THRESHOLD) {
                                return currentPosition + 1;
                            } else {
                                return currentPosition;
                            }
                        } else {
                            throw new IllegalArgumentException("랜덤 값은 0~9 사이여야 합니다.");
                        }
                    } else {
                        throw new IllegalArgumentException("랜덤 값은 0~9 사이여야 합니다.");
                    }
                } else {
                    throw new IllegalArgumentException("자동차 이름은 5자를 초과할 수 없습니다.");
                }
            } else {
                throw new IllegalArgumentException("자동차 이름은 비어있을 수 없습니다.");
            }
        } else {
            throw new IllegalArgumentException("자동차 이름은 null일 수 없습니다.");
        }
    }
}

File: ts/step2-early-return/deepRaceValidator.ts
export class DeepRaceValidator {
    private static readonly MAX_NAME_LENGTH = 5;
    private static readonly MOVE_THRESHOLD = 4;
    private static readonly MAX_RANDOM_VALUE = 9;
    private static readonly MIN_RANDOM_VALUE = 0;

    public move(
        name: string | null,
        randomValue: number,
        currentPosition: number
    ): number {


        if (name === null) {
            throw new Error('자동차 이름은 null일 수 없습니다.');
        }

        if (name.length === 0) {
            throw new Error('자동차 이름은 비어있을 수 없습니다.');
        }

        if (name.length > DeepRaceValidator.MAX_NAME_LENGTH) {
            throw new Error('자동차 이름은 5자를 초과할 수 없습니다.');
        }


        if (randomValue < DeepRaceValidator.MIN_RANDOM_VALUE ||
            randomValue > DeepRaceValidator.MAX_RANDOM_VALUE) {
            throw new Error('랜덤 값은 0~9 사이여야 합니다.');
        }

        
        if (randomValue < DeepRaceValidator.MOVE_THRESHOLD) {
            return currentPosition;
        }

        return currentPosition + 1;
    }
}